site/react/App.jsx

```diff
@@ -0,0 +1,62 @@
+import React, { useState, useEffect } from 'react'
+
+import { listFrameworks } from '../../dist'
+
+const repo = 'netlify-templates/gatsby-starter-netlify-cms'
+
+const get = async (path) => {
+  const url = path ? `https://api.github.com/repos/${repo}/${path}` : `https://api.github.com/repos/${repo}`
+  const response = await fetch(url)
+  return response.json()
+}
+
+const listFiles = async () => {
+  const { default_branch } = await get()
+  const { tree } = await get(`git/trees/${default_branch}?recursive=1`)
+  return tree.map(({ path }) => path)
+}
+
+const getPackageJson = async () => {
+  const { content } = await get('contents/package.json')
+  const raw = atob(content)
+  return { packageJson: JSON.parse(raw), packageJsonPath: '' }
+}
+
+const getContext = async () => {
+  const repoFiles = await listFiles()
+  const locatePath = (paths) => {
+    const foundPath = paths.find((path) => {
+      const normalizedPath = path.startsWith('./') ? path.slice(2) : path
+      return repoFiles.includes(normalizedPath)
+    })
+    return foundPath
+  }
+  return { projectDir: '.', locatePath, getPackageJson }
+}
+
+const Framework = (framework) => {
+  return <div key={framework.name}>{JSON.stringify(framework)}</div>
+}
+
+export const App = () => {
+  const [frameworks, setFrameworks] = useState([])
+
+  useEffect(() => {
+    let canceled = false
+
+    const fetchFrameworks = async () => {
+      const context = await getContext()
+      const frameworks = await listFrameworks(context)
```

Comment:
Should we provide two default contexts?
  - One for Node.js, using the filesystem
  - One for browsers, using an array of files

Like this, most consumers would only need to specify which of those two contexts they want to use, instead of having to implement `locatePath()` and `getPackageJson()`?

Reply:
In some cases it's better to provide another implementation for `locatePath` (e.g for the browser listing files can be an expensive operation depending on the `git` service - with GitLab you'd have to handle pagination instead of testing if a file exists using a `HEAD` request).

Reply:
Should we provide a default implementation though for the most common use cases (especially for browser consumers)?

Reply:
> Should we provide a default implementation though for the most common use cases

I added another [commit](https://github.com/netlify/framework-info/pull/58/commits/b6c655d844a156c74c33b95e147d9bd3f903c52c) to simplify `context` so it only requires a `pathExists` method and a `packageJson`.
Does that make things simpler? 

I believe the only `git` service that benefits from using a file list is GitHub as at the moment.
How about we add a default context that accepts a file list if we see that other services benefit from it?

Also, there is another (minor) improvement I would like to do - we can check for `yarn.lock` (to evaluate `runScriptCommand`) only if we detected a framework, saving that network request (which would be done every time at the moment). I think it's better to do it in another PR

Reply:
Thanks!
So, if I understand correctly, you're saying that GitLab and Bitbucket would need separate logic than GitHub, is that right?

My main question is where the implementation of those different providers should be. We could ask our front-end to implement all of those. However, it seems to me other consumers might benefit from this as well, so it might make sense to implement it this library instead? This was what I meant by "providing a default implementation". What do you think?

Reply:
> So, if I understand correctly, you're saying that GitLab and Bitbucket would need separate logic than GitHub, is that right?

Oh, I get what you mean now. Yes I believe Bitbucket and GitLab would have a different implementation (instead of listing files and using `includes` they would try to directly retrieve the path to check for existence).

Is your suggestion to have the logic to access GitHub/Bitbucket/GitLab as a part of this library and the consumer pass in a token + provider name?

I don't think specific providers logic should be a part of this lib.
To clarify what that logic would like for GitLab you can have a look at:
https://github.com/netlify/netlify-cms/blob/8b76c2388832580f73abf4c414333452e8c915f9/packages/netlify-cms-backend-azure/src/API.ts#L600
That code can be simplified as we will only be checking the default branch, so the logic will just a few lines of code to send a single request and handle the response.

Reply:
> Is your suggestion to have the logic to access GitHub/Bitbucket/GitLab as a part of this library and the consumer pass in a token + provider name?

Yes, that's exactly what I meant. With the additional possibility to implement your own provider if none of the existing ones fit the needs.

> I don't think specific providers logic should be a part of this lib.

Where do you suggest the logic should be? In each consumer, or in a separate library?

Reply:
I'll Bitbucket and GitLab to the example site

Reply:
> Where do you suggest the logic should be? In each consumer, or in a separate library?

I think it should be quite small so each consumer can handle it, considering you'd have to pass an auth token and sometimes even an API endpoint (e.g. GitHub enterprise).
A library (at the moment) would just be a single fetch request per provider.

Since we only have a single browser consumer at the moment I'm good with having that logic in the consumer level.
If we see it's being duplicated we can add it.

Reply:
If this is open-source, do you think other users might want to use this library, i.e. we might have other consumers who would need to do the same?
I am fine with this solution, just wondering.

Reply:
> If this is open-source, do you think other users might want to use this library, i.e. we might have other consumers who would need to do the same?

I think it should be easier to reason on how to share that logic when we have more than one browser consumer.

Reply:
We have been discussing having more parts of the company using this library, and how this library might fit in our general framework-agnostic yet framework-friendly strategy. From that standpoint, I think it would be nice to make this library as easy to consume as possible and shift the complexity from the consumer to the core instead.

I do feel like we might be adding some implementation obstacles for consumers of this library with this design, where we might not want to. However, this PR does solve our more immediate problem of making it accessible to front-end, so if this direction sounds good to you, this sounds good to me too :+1:

Reply:
Opened https://github.com/netlify/framework-info/issues/68